feature: Add DND button in left sidebar.

jainaman224 opened the pull request:
Fixes: #298 

**What's this PR do?**
This adds the do not disturb button to the left sidebar.

**Screenshots?**
![ezgif com-video-to-gif 2](https://user-images.githubusercontent.com/13117482/36254026-8e42bb76-126f-11e8-99a8-faf0094515a7.gif)

**You have tested this PR on:**
  - [ ] Windows
  - [x] Linux/Ubuntu
  - [ ] macOS

priyank-p commented:
@jainaman224 this PR is totally broken the button still does't work have you tested this?

jainaman224 commented:
@cPhost Yes, it is working for me.

jainaman224 commented:
@cPhost What were the initial settings of `Show Notification` and `Mute all sounds`?

priyank-p commented on app/main/menu.js:
```diff
@@ -366,9 +378,19 @@ class AppMenu {
 	}
 
+	static doNotDisturb() {
+		const newValue = !((ConfigUtil.getConfigItem('showNotification', true) ||
```
I don't think this is the right approch here the whole conditional on one line does seem like a good idea its very bug prone. 

Can you instead put the list of all the setting in and array like `var settingList  = ['showNotification', ..]` and then loop through it using `.map` in one liner? 
(Do let me if this is unclear)

priyank-p commented on app/renderer/js/main.js:
```diff
@@ -152,6 +154,35 @@ class ServerManagerView {
 	initActions() {
+		this.initDNDButton();
+		this.$dndButton.addEventListener('click', () => {
+			const dnd = ConfigUtil.getConfigItem('dnd', false);
```
Once again this is very hard to review an trace back the where the bug is happing if the situation arises. This needs a lot of code refactoring. And this might explain why is not working for me locally.

priyank-p reviewed (changes requested):
@jainaman224 thanks for working on it, but i think the code need to better refactored and this could be made very simple. Currently is very hard to review and know whats going on and why? I do get the points of function but you can do better here.

priyank-p commented:
@jainaman224 the weird errors we were facing, were because you did not add `dnd` key in `ConfigUtil` so you can't get retrive it hence causing all the errors. You need to add default setting here https://github.com/zulip/zulip-electron/blob/master/app/renderer/js/main.js#L77

priyank-p commented on app/renderer/js/main.js:
```diff
@@ -87,7 +89,14 @@ class ServerManagerView {
+			oldSettings: {
+				showNotification: false,
+				silent: true,
+				flashTaskbarOnMessage: false
```
`oldSettings` by default should be same as default.

priyank-p commented on app/main/menu.js:
```diff
@@ -399,6 +411,11 @@ class AppMenu {
 		});
 	}
 
+	static doNotDisturb() {
```
Can you move this into `click` handler above

priyank-p commented on app/main/menu.js:
```diff
@@ -211,6 +211,26 @@ class AppMenu {
+						oldSettings = {};
+						for (const settingName of settingList) {
+							currentSettings[settingName] = ConfigUtil.getConfigItem(settingName);
+							oldSettings[settingName] = (settingName === 'silent');
```
This right here seems bit hacky, so i'd add a comment here to explain what's going on for future maintaing.

priyank-p reviewed (commented):
So for functionality wise looks good @jainaman224! But can you move the handling of the dnd into its own util at `js/utils`. And also a minor comment.

jainaman224 commented:
@cPhost Please review.

priyank-p commented on app/renderer/js/utils/dnd-util.js:
```diff
@@ -0,0 +1,34 @@
+
+const ConfigUtil = require(__dirname + '/config-util.js');
+
+function dnd() {
```
`dnd` -> `toogle` so `DNDUtil.toogle`

priyank-p commented on app/renderer/js/utils/dnd-util.js:
```diff
@@ -0,0 +1,34 @@
+const ConfigUtil = require(__dirname + '/config-util.js');
+
+function dnd() {
+	this.initDNDButton();
```
This is from the old commit which should be removed, or i'd think you do need to implement the old thing here>

jainaman224 replied:
I think, we need to initialize the button.

priyank-p replied:
Yeah, so you need to do this here :)

priyank-p commented on app/renderer/js/utils/dnd-util.js:
```diff
@@ -0,0 +1,34 @@
+	this.initDNDButton();
+	const dnd = !ConfigUtil.getConfigItem('dnd', false);
+	const settingList = ['showNotification', 'silent', 'flashTaskbarOnMessage'];
+	let oldSettings;
```
Remove this from here and initialize variable inside the `if`/`else` statement.

jainaman224 replied:
I have to initialize it outside if/else. Due to scoping issue. It is inaccessible outside the block if I initialize it inside the conditional block.

priyank-p replied:
Ignore this, i was think of change of variable and stuff :)

priyank-p commented on app/renderer/js/utils/dnd-util.js:
```diff
@@ -0,0 +1,34 @@
+	} else {
+		oldSettings = ConfigUtil.getConfigItem('oldSettings');
+	}
+	for (const settingName of settingList) {
```
new line here!

priyank-p commented on app/renderer/js/utils/dnd-util.js:
```diff
@@ -0,0 +1,34 @@
+	for (const settingName of settingList) {
+		ConfigUtil.setConfigItem(settingName, oldSettings[settingName]);
+	}
+	ConfigUtil.setConfigItem('dnd', dnd);
```
Ditto!

priyank-p commented on app/renderer/js/utils/dnd-util.js:
```diff
@@ -0,0 +1,34 @@
+	let oldSettings;
+	if (dnd) {
+		const currentSettings = {};
+		oldSettings = {};
```
I'd do change the `currentSetting` -> `oldSetting` and `oldSettings` to `newSettings` or `dndSetting`.  They are used in way seems confusing.

priyank-p reviewed (commented):
Left a bunch of comments again

jainaman224 commented:
@akashnimare Please review.

akashnimare commented on app/renderer/js/utils/dnd-util.js:
```diff
@@ -0,0 +1,43 @@
+
+function toggle() {
+	const dnd = !ConfigUtil.getConfigItem('dnd', false);
+	let settingList;
```
We can re-write this as - 
```javascript
const settingList = ['showNotification', 'silent'];
if (process.platform === 'win32') {
    settingList.push('flashTaskbarOnMessage')
}
```
maybe rename `settingList` to `dndSettingList`, cc @cPhost.

akashnimare commented on app/renderer/js/main.js:
```diff
@@ -87,7 +90,12 @@ class ServerManagerView {
+			lastActiveTab: 0,
+			dnd: false,
+			oldSettings: {
+				showNotification: true,
```
I think this should inherit from the main settings, if it's not null. Something like this - 
```javascript
oldSettings: {
	showNotification: ConfigUtil.getConfigItem('showNotification') || true,
	silent: ConfigUtil.getConfigItem('silent') || false
}

```

jainaman224 replied:
@akashnimare What are the advantages?

I think it should be like this.
```js
oldSettings: {
	showNotification: (ConfigUtil.getConfigItem('showNotification') === null) || true,
	silent: (ConfigUtil.getConfigItem('silent') === null) || false
}
```

priyank-p commented on app/renderer/js/utils/dnd-util.js:
```diff
@@ -0,0 +1,43 @@
+function toggle() {
+	const dnd = !ConfigUtil.getConfigItem('dnd', false);
+	let dndSettingList;
+	if (process.platform === 'win32') {
```
This if else block can be simplified to be
```javascript
const dndSettingList = ['showNotification', 'silent'];
if (process.platfrom == 'win32') {
  dndSettingList.push('flashTaskbarOnMessage');
}
```

priyank-p commented on app/renderer/js/utils/dnd-util.js:
```diff
@@ -0,0 +1,43 @@
+		dndSettingList = ['showNotification', 'silent'];
+	}
+	let newSettings;
+
```
Remove the space and add it above!

priyank-p reviewed (commented):
Can you fix this basic styling issues with your code thanks!

akashnimare commented on app/renderer/js/main.js:
```diff
@@ -87,7 +90,12 @@ class ServerManagerView {
-			lastActiveTab: 0
+			lastActiveTab: 0,
+			dnd: false,
+			oldSettings: {
```
@cPhost, @jainaman224 can we make it more readable? I was thinking to use something like "dndState", "dndOldState", "dndPreviousSettings".

priyank-p replied:
Yeah, we can this make's it more clear. (Not that it was not clear before, but yeah this is great)

jainaman224 replied:
I have changed it to `dndPreviousSettings`.

jainaman224 commented on app/renderer/js/main.js:
```diff
@@ -87,7 +90,12 @@ class ServerManagerView {
+			lastActiveTab: 0,
+			dnd: false,
+			oldSettings: {
+				showNotification: ConfigUtil.getConfigItem('showNotification') || true,
```
@cPhost @akashnimare This statement gives true always. So, we should go for conditional statement.

```js
showNotification: ConfigUtil.getConfigItem('showNotification') === null ? true : ConfigUtil.getConfigItem('showNotification'),
```

priyank-p replied:
It shouldn't return true if its disabled in settings. It returns `true` since the default it true.

priyank-p replied:
To be clear this doesn't need a check!

priyank-p replied:
Err! The way it is is fine.

akashnimare commented:
@jainaman224 great work so for. We just need to add two more things here -
* Document the shortcuts (CTRL+SHIFT+M) on our shortcuts page.
* Update the content of tooltip depending upon the state of DND.

jainaman224 commented:
@akashnimare I have made required changes. Please review this PR.

jainaman224 commented:
@abhigyank It is used for the maintaining similarity across codebase. Have a look at https://github.com/zulip/zulip-electron/blob/master/app/renderer/js/main.js#L307.

priyank-p commented:
@jainaman224 the commit-msg needs to be fixed, `feature` instead of `design` and you need to remove the testing description from commit msg, generally, you should be describing what the commit is about unless it a test.

jainaman224 commented:
@cPhost @akashnimare I have made required changes. Please review.

jainaman224 commented:
@cPhost Made the required changes. Please review.

priyank-p commented:
@jainaman224 thanks! @akashnimare can you cross-check that this fixes the flashing on windows.

akashnimare commented:
@jainaman224 there is something wrong with the DND functionality. The settings (Desktop notifications, silent) are changing automatically if you change it and then switch back to a server. 

![dnd](https://user-images.githubusercontent.com/2263909/37529660-b316396e-295d-11e8-9874-f16c15d704d5.gif)

akashnimare commented:
I can reproduce this bug in the master as well.

abhigyank commented:
I test this on Ubuntu 16.04. 
This doesn't work for me.  
1. I think  "Show Desktop Notifications" is broken in the master itself. Turning off "Show Desktop Notifications" still gets me notifications even on the master.
2. The DND button's functionality doesn't work for me anyway. It changes the settings for me alright in the settings window, but doesn't changes the functionality. It still works on the previous functionality. 
With "Mute all sounds" off initially, turning DND on, I still get sound, while with "Mute all sounds" on initially, turning DND on, I don't get sound. So it seems, it shows the previous behavior. 
3. I think if someone manually changes the settings to the DND position, the DND button should change its icon to mute.
Can others replicate this?

jainaman224 commented:
Point 1, it was fixed by Akash as far as I know.
Point 3, It was discussed with team and we are planning to work on it next.
Point 2, I will test it again. Thanks for testing

akashnimare commented:
@jainaman224 can you rebase and fix the merge conflicts?

jainaman224 committed:
feature: Add DND button in left sidebar.

This adds the do not disturb button to the left sidebar
which disables sound and notifications. It also disables
flash taskbar on windows.
Fixes: #298

akashnimare commented:
This is what I'm experiencing on macOS -
* Fresh install + added `czo` server
* Clicking the DND button activates the DND mode but I still get the sound no notification though
* After hard reloading, it works as expected (no notification and no sound)
* To disable the DND mode I still need to hard reload otherwise (notification + no sound)

abhigyank commented:
I did a Fresh install + added 'czo' server too.
* No longer having the "Show Desktop Notifications" issue.
* I can corroborate @akashnimare's experience on Ubuntu 16.04 too.
* Adding to it, with "Mute all sounds" off initially, turning DND on, I still get sound (no notification though), while with "Mute all sounds" on initially, turning DND on, I don't get sound (no notification). So it seems, it shows the previous behavior for sound.
* Same with disable. Hard reload solves both issues.

akashnimare commented:
Merged in https://github.com/zulip/zulip-electron/commit/537fbe8f9e0f0dbb573ac4ffe140c0a580d17f24 after minor change.
